client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -60,6 +62,10 @@
 
     // Base waiting time for a reader on an idle segment waiting for new data to be read.
     private static final long BASE_READER_WAITING_TIME_MS = ReaderGroupStateManager.TIME_UNIT.toMillis();
+    // As an optimization to do not generate Position objects on every event read, we define a base map of segments and
+    // then a batch of updates to the offsets of these segments, one per event read. Internally, the Position object can
+    // derive the right offsets at which the event was read by lazily replying such updates up to the point it was read.
+    private static final long MAX_BUFFERED_SEGMENT_OFFSET_UPDATES = 1000;
```

Comment:
Because a copy is needed anyway (see comments below) this is actually going to make things slower. It would be better to simply in the `getPosition()` method iterate over the `readers` and construct the `List<Entry<Segment, Long>>` with only one entry per-segment. This is faster, avoids both of the need collections here, and the need to try to cache the position object.

Reply:
I have added a performance comparison comparing your suggestion with this PR. For 1 segment, your approach is at par with this PR in terms of speed. But as more segments are handled by the reader, the performance decreases, given that the list to copy is larger.